### ISSUE TITLE
Exception with ELF MIPSII

### ISSUE THREAD

**Issue by g05u** [pwndbg/pwndbg#156]
Hey, when I set a breakpoint in an ELF MIPSII using remote debugging, pwndbg throws an exception (binary is attached):

[bender.gz](https://github.com/pwndbg/pwndbg/files/742558/bender.gz)

pwndbg> file bender_safe 
Reading symbols from bender_safe...(no debugging symbols found)...done.
pwndbg> b validate 
Breakpoint 1 at 0x401c84
pwndbg> c
Continuing.
Traceback (most recent call last):
  File "/home/user/pwndbg/pwndbg/events.py", line 111, in caller
    func()
  File "/home/user/pwndbg/pwndbg/vmmap.py", line 107, in explore_registers
    find(pwndbg.regs[regname])
...
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
Traceback (most recent call last):

**Comment by disconnect3d** [comment]
This is a remote debugging session, right?

You probably have to set proper architecture, e.g.:
```
set arch mips
```
you must have the gdb installed with this architecture, or you can try gdb-multiarch.

However I am not sure the pwndbg works with MIPS.

**Comment by g05u** [comment]
When I have a remotely debuged process and I try to read the local symbols (file path/to/bin) pwndbg doesn't manage it successfully and an exception is thrown.

**Comment by zachriggle** [comment]
The issue is that the internal ELF structures declared via `ctypes` parse the data from the ELF files in-memory in Little Endian format, despite the actual contents being Big Endian.

There's some hackery we can probably do with metaclasses or __getattribute__ hooks.

**Comment by zachriggle** [comment]
I've added some code to the `dev` branch which prevents all of these error messages.

However, it looks like Capstone's disassembler is broken for this architecture.

**Comment by zachriggle** [comment]
Filed aquynh/capstone#849 and added a work-around to `dev`

**Comment by zachriggle** [comment]
Looks like most things work.  The branch detection via Unicorn is busted, but that's it.
<img width="1039" alt="screen shot 2017-01-31 at 9 58 40 pm" src="https://cloud.githubusercontent.com/assets/111640/22493831/7983bfae-e800-11e6-917e-665bde7b3523.png">

**Comment by zachriggle** [comment]
Merged to `master`

**Comment by g05u** [comment]
Thanks, this issue is fixed

**Comment by disconnect3d** [comment]
Is this really fixed?

I still have those memory errors. I launch the binary in one console by:
```
$ ./qemu-mips -g 7777 ./bender_safe
```

And then in another one:
```
[dc@dc:challenge]$ gdb ./bender_safe --quiet 
Loaded 104 commands.  Type pwndbg [filter] for a list.
Reading symbols from ./bender_safe...(no debugging symbols found)...done.
pwndbg> set arch mips
The target architecture is assumed to be mips
pwndbg> target remote localhost:7777
Remote debugging using localhost:7777
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 120, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/stack.py", line 79, in update
    page  = pwndbg.memory.Page(start, stop-start, 6 if not is_executable() else 7, 0, '[stack]')
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/stack.py", line 127, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 176, in iter_phdrs
    phnum, phentsize, phdr = get_phdrs(ehdr.address)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 169, in get_phdrs
    x = (phnum, phentsize, read(Phdr, Elfhdr.address + phoff))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 59, in read
    data = pwndbg.memory.read(address, size)
  File "/home/dc/installed/pwndbg/pwndbg/memory.py", line 43, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 120, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/stack.py", line 127, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 176, in iter_phdrs
    phnum, phentsize, phdr = get_phdrs(ehdr.address)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 169, in get_phdrs
    x = (phnum, phentsize, read(Phdr, Elfhdr.address + phoff))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 59, in read
    data = pwndbg.memory.read(address, size)
  File "/home/dc/installed/pwndbg/pwndbg/memory.py", line 43, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 120, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 107, in explore_registers
    find(pwndbg.regs[regname])
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 63, in find
    for page in get():
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 47, in get
    else:     pages.extend(info_files())
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 290, in info_files
    pages.extend(pwndbg.elf.map(vaddr, objfile))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 210, in map
    return map_inner(ei_class, ehdr, objfile)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 226, in map_inner
    for phdr in iter_phdrs(ehdr):
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 176, in iter_phdrs
    phnum, phentsize, phdr = get_phdrs(ehdr.address)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 169, in get_phdrs
    x = (phnum, phentsize, read(Phdr, Elfhdr.address + phoff))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 59, in read
    data = pwndbg.memory.read(address, size)
  File "/home/dc/installed/pwndbg/pwndbg/memory.py", line 43, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 120, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/symbol.py", line 244, in add_main_exe_to_symbols
    mmap = pwndbg.vmmap.find(addr)
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 63, in find
    for page in get():
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 47, in get
    else:     pages.extend(info_files())
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/vmmap.py", line 290, in info_files
    pages.extend(pwndbg.elf.map(vaddr, objfile))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 210, in map
    return map_inner(ei_class, ehdr, objfile)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 226, in map_inner
    for phdr in iter_phdrs(ehdr):
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 176, in iter_phdrs
    phnum, phentsize, phdr = get_phdrs(ehdr.address)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 169, in get_phdrs
    x = (phnum, phentsize, read(Phdr, Elfhdr.address + phoff))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 59, in read
    data = pwndbg.memory.read(address, size)
  File "/home/dc/installed/pwndbg/pwndbg/memory.py", line 43, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
0x00401160 in __start ()
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/pwndbg/events.py", line 120, in caller
    func()
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/stack.py", line 79, in update
    page  = pwndbg.memory.Page(start, stop-start, 6 if not is_executable() else 7, 0, '[stack]')
  File "/home/dc/installed/pwndbg/pwndbg/memoize.py", line 47, in __call__
    value = self.func(*args, **kwargs)
  File "/home/dc/installed/pwndbg/pwndbg/stack.py", line 127, in is_executable
    for phdr in pwndbg.elf.iter_phdrs(ehdr):
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 176, in iter_phdrs
    phnum, phentsize, phdr = get_phdrs(ehdr.address)
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 169, in get_phdrs
    x = (phnum, phentsize, read(Phdr, Elfhdr.address + phoff))
  File "/home/dc/installed/pwndbg/pwndbg/elf.py", line 59, in read
    data = pwndbg.memory.read(address, size)
  File "/home/dc/installed/pwndbg/pwndbg/memory.py", line 43, in read
    result = gdb.selected_inferior().read_memory(addr, count)
gdb.MemoryError: Cannot access memory at address 0x34400000
Python Exception <class 'gdb.MemoryError'> Cannot access memory at address 0x34400000: 
```

I am using Arch Linux if that matters, also:
```
[dc@dc:pwndbg|master]$ gdb --version
GNU gdb (GDB) 7.12
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
[dc@dc:pwndbg|master]$ git show HEAD
commit 064a2c1ea84a65c30592f7d6ef8f693097454603
Author: Zach Riggle <riggle@google.com>
Date:   Mon Mar 6 11:37:50 2017 -0600

    Add docs requirements file
```

**Comment by zachriggle** [comment]
Does adfa644 work for you, @disconnect3d

**Comment by zachriggle** [comment]
In your GDB script you didn't "set endian big"

**Comment by disconnect3d** [comment]
The commit adfa644 works for me  (even without `set endian big`).
The latest commit on master doesn't work even when I set endianness.

I have made a `git bisect` and found out the "first bad commit" as 8110dacacbc69ba14e22aa2a49c7bc22b5348fe2.
```
[dc@dc:pwndbg|(5f0e842...)|BISECTING]$ git bisect good
8110dacacbc69ba14e22aa2a49c7bc22b5348fe2 is the first bad commit
commit 8110dacacbc69ba14e22aa2a49c7bc22b5348fe2
Author: g05u <henrybytes@gmail.com>
Date:   Wed Feb 1 22:29:21 2017 -0500

    Fix disassembler extra=None vs extra=0 (#157)
```

However reverting this commit on master didn't fix the issue. I tried looking more, by bisecting with temporary modifications (reverting and reseting the "bad commit") and found out the problem stays there just for the next commit - 9d19daaea3815066ea0b4fa83bfb30c8bf5f260f.

**Comment by disconnect3d** [comment]
In [elf.py the `update()`](https://github.com/pwndbg/pwndbg/blob/master/pwndbg/elf.py#L42) sets `pwndbg.elf.Ehdr` to `pwndbg.elftypes.ElfXX_Ehdr`.
The `pwndb.elftypes.ElfXX_Ehdr` inherits from `pwndbg.ctypes.Structure`.

When initializing pwndbg, the `pwndbg.arch.endian` is set to `little` and so the `pwndbg.ctypes` sets its `Structure` attribute to `LittleEndianStructure`. Because of that the `pwndbg.elftypes.ElfXX_Ehdr` already inherits from `Structure` which is `LittleEndianStructure`.

There is the [`update` method in `pwndbg.ctypes`](https://github.com/pwndbg/pwndbg/blob/master/pwndbg/ctypes.py#L25) which is meant to set proper endianness structure in particular times but it isn't really helpful at all. It changes the `pwndbg.ctypes.Structure` to proper value, but the  `pwndbg.elftypes.ElfXX_Ehdr` has already been set and inherits from the old `pwndbg.ctypes.Structure` which is `LittleEndianStructure`.

Setting endianness to `big` in .gdbinit (`set endian big`) before initializing pwndbg doesn't help because the `pwndbg.arch.endian` isn't really a gdb parameter (or at least I believe so) and [is set to `little` in pwndbg/arch.py](https://github.com/pwndbg/pwndbg/blob/master/pwndbg/arch.py#L23).

Setting `endian = 'big'` in pwndbg/arch.py seems to be a workaround for the problem.

**Comment by zachriggle** [comment]
Thanks for tracking this down, @disconnect3d.  I presume that some of the changes to `pwndbg.ctypes` might have caused this regression.  At some point in time we had code that would re-load that entire module at run-time, so that the structures were re-created with the correct endianness -- but it looks like that was removed in favor of a different update mechanism.  Not entirely sure why I did this.
